File: packages/core/src/pft2/legacy.test.ts
import { describe, expect, it } from "vitest";
import { Pft2LegacyBaseTree, type Pft2LegacyManifestEntry } from "./basetree.js";
import {
  Pft2CorruptError,
  Pft2FileKind,
  Pft2InvalidNodeError,
  Pft2NotFoundError,
  checkNodeRefBounds,
} from "./types.js";

function legacyManifestEntries(): Pft2LegacyManifestEntry[] {
  return [
    {
      path: "src/main.go",
      kind: "file",
      mode: 0o644,
      size: 9000000,
      blob: { digest: "sha256:aaaa", size: 9000000 },
      chunks: [
        { digest: "sha256:c1", size: 4194304, offset: 0 },
        { digest: "sha256:c2", size: 4194304, offset: 4194304 },
        { digest: "sha256:c3", size: 611392, offset: 8388608 },
      ],
      ino: 77n,
    },
    {
      path: "README.md",
      kind: "file",
      mode: 0o644,
      size: 5,
      blob: { digest: "sha256:bbbb", size: 5 },
      mtimeMs: 1700000000000,
    },
    { path: "src", kind: "directory", mode: 0o755 },
    { path: "docs/deep/nested.txt", kind: "file", mode: 0o600, size: 0 },
    { path: "link", kind: "symlink", mode: 0o777, linkTarget: "README.md" },
  ];
}

describe("Pft2LegacyBaseTree", () => {
  it("adapts a legacy manifest with synthesized directories and stable inos", async () => {
    const tree = new Pft2LegacyBaseTree(legacyManifestEntries());
    const root = await tree.getInode(1n);
    expect(root.inode.kind).toBe(Pft2FileKind.Directory);

    const srcEntry = await tree.lookup(root.ref, "src");
    const src = await tree.getInode(srcEntry.ino);
    const mainEntry = await tree.lookup(src.ref, "main.go");
    expect(mainEntry.ino).toBe(77n);

    const docsEntry = await tree.lookup(root.ref, "docs");
    expect(docsEntry.kind).toBe(Pft2FileKind.Directory);
    expect(docsEntry.ino > 77n).toBe(true);
    const docs = await tree.getInode(docsEntry.ino);
    const deepEntry = await tree.lookup(docs.ref, "deep");
    const deep = await tree.getInode(deepEntry.ino);
    await expect(tree.lookup(deep.ref, "nested.txt")).resolves.toBeDefined();

    const names: string[] = [];
    let cursor = "";
    for (;;) {
      const { entries, next } = await tree.readDir(root.ref, cursor, 2);
      names.push(...entries.map((entry) => entry.name));
      if (next === "") {
        break;
      }
      cursor = next;
    }
    expect(names).toEqual(["README.md", "docs", "link", "src"]);

    const linkEntry = await tree.lookup(root.ref, "link");
    const link = await tree.getInode(linkEntry.ino);
    expect(link.inode.symlinkTarget).toBe("README.md");
    expect(link.inode.size).toBe(BigInt("README.md".length));

    await expect(tree.lookup(root.ref, "absent")).rejects.toThrow(Pft2NotFoundError);
    await expect(tree.getInode(123456n)).rejects.toThrow(Pft2NotFoundError);
  });

  it("exposes legacy blob and chunk objects as extents", async () => {
    const tree = new Pft2LegacyBaseTree(legacyManifestEntries());
    const main = await tree.getInode(77n);

    const spanning = await tree.readExtents(main.ref, 4194304n - 100n, 200n);
    expect(spanning).toHaveLength(2);
    expect(spanning[0]!.legacy!.objectDigest).toBe("sha256:c1");
    expect(spanning[0]!.fileOffset).toBe(4194304n - 100n);
    expect(spanning[0]!.length).toBe(100n);
    expect(spanning[0]!.legacy!.objectOffset).toBe(4194304n - 100n);
    expect(spanning[1]!.legacy!.objectDigest).toBe("sha256:c2");
    expect(spanning[1]!.fileOffset).toBe(4194304n);
    expect(spanning[1]!.length).toBe(100n);
    expect(spanning[1]!.legacy!.objectOffset).toBe(0n);

    const root = await tree.getInode(1n);
    const readmeEntry = await tree.lookup(root.ref, "README.md");
    const readme = await tree.getInode(readmeEntry.ino);
    const window = await tree.readExtents(readme.ref, 1n, 100n);
    expect(window).toHaveLength(1);
    expect(window[0]!.fileOffset).toBe(1n);
    expect(window[0]!.length).toBe(4n);
    expect(window[0]!.legacy!.objectOffset).toBe(1n);
    expect(window[0]!.legacy!.objectSize).toBe(5n);

    expect(await tree.readExtents(main.ref, 9000000n, 10n)).toHaveLength(0);
    await expect(tree.readExtents(root.ref, 0n, 10n)).rejects.toThrow(Pft2CorruptError);
  });

  it("rejects malformed legacy manifests", () => {
    const cases: Record<string, Pft2LegacyManifestEntry[]> = {
      "duplicate ino": [
        { path: "a", kind: "file", mode: 0o644, ino: 5n },
        { path: "b", kind: "file", mode: 0o644, ino: 5n },
      ],
      "root ino claimed": [{ path: "a", kind: "file", mode: 0o644, ino: 1n }],
      "duplicate path": [
        { path: "a", kind: "file", mode: 0o644 },
        { path: "a", kind: "directory", mode: 0o755 },
      ],
      "file as parent": [
        { path: "a", kind: "file", mode: 0o644 },
        { path: "a/b", kind: "file", mode: 0o644 },
      ],
      "blob size mismatch": [
        { path: "a", kind: "file", mode: 0o644, size: 10, blob: { digest: "sha256:x", size: 9 } },
      ],
      "missing blob": [{ path: "a", kind: "file", mode: 0o644, size: 10 }],
      "non-contiguous chunks": [
        {
          path: "a",
          kind: "file",
          mode: 0o644,
          size: 10,
          chunks: [{ digest: "sha256:c", size: 5, offset: 1 }],
        },
      ],
      "chunk sum mismatch": [
        {
          path: "a",
          kind: "file",
          mode: 0o644,
          size: 10,
          chunks: [{ digest: "sha256:c", size: 5, offset: 0 }],
        },
      ],
      "bad path": [{ path: "a//b", kind: "file", mode: 0o644 }],
      "dotdot path": [{ path: "a/../b", kind: "file", mode: 0o644 }],
      "negative size": [{ path: "a", kind: "file", mode: 0o644, size: -1 }],
      // Full inode-invariant parity with the Go adapter (gate F): every
      // synthesized inode passes the exact PFT2 validation rules.
      "empty symlink target": [{ path: "l", kind: "symlink", mode: 0o777, linkTarget: "" }],
      "symlink target with NUL": [{ path: "l", kind: "symlink", mode: 0o777, linkTarget: "a\0b" }],
      "timestamp beyond bound": [
        { path: "a", kind: "file", mode: 0o644, mtimeMs: 2 ** 60 },
      ],
      "negative timestamp beyond bound": [
        { path: "a", kind: "file", mode: 0o644, ctimeMs: -(2 ** 60) },
      ],
      "uid out of range": [{ path: "a", kind: "file", mode: 0o644, uid: 2 ** 40 }],
      "gid not an integer": [{ path: "a", kind: "file", mode: 0o644, gid: 1.5 }],
    };
    for (const [name, entries] of Object.entries(cases)) {
      expect(() => new Pft2LegacyBaseTree(entries), name).toThrow(Pft2InvalidNodeError);
    }
  });

  it("masks legacy modes to the stored permission bits like Go", async () => {
    // Both adapters keep only the 0o7777 mode bits (S_IFMT bits are carried
    // by the kind), so a legacy mode with type bits still adapts.
    const tree = new Pft2LegacyBaseTree([
      { path: "a", kind: "file", mode: 0o100644, size: 0 },
    ]);
    const root = await tree.getInode(1n);
    const entry = await tree.lookup(root.ref, "a");
    const file = await tree.getInode(entry.ino);
    expect(file.inode.mode).toBe(0o644);
  });

  it("issues handles that fail closed against real PFT2 readers", async () => {
    const tree = new Pft2LegacyBaseTree(legacyManifestEntries());
    const root = await tree.getInode(1n);
    // Legacy handles have size 0, below the minimum node size, so they can
    // never pass the pre-fetch bounds check of a real reader.
    expect(() => checkNodeRefBounds("probe", root.ref)).toThrow(Pft2InvalidNodeError);
  });
});

File: vcs/internal/pft2/legacy_test.go
package pft2

import (
	"context"
	"errors"
	"testing"
)

func legacyManifestEntries() []LegacyEntry {
	return []LegacyEntry{
		// Deliberately unordered; adapter sorts by path bytes.
		{Path: "src/main.go", Kind: "file", Mode: 0o644, Size: 9000000,
			BlobDigest: "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			BlobSize:   9000000,
			Chunks: []LegacyChunk{
				{Digest: "sha256:c1", Size: 4194304, Offset: 0},
				{Digest: "sha256:c2", Size: 4194304, Offset: 4194304},
				{Digest: "sha256:c3", Size: 611392, Offset: 8388608},
			},
			Ino: 77},
		{Path: "README.md", Kind: "file", Mode: 0o644, Size: 5,
			BlobDigest: "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
			BlobSize:   5, MtimeMs: 1700000000000},
		{Path: "src", Kind: "directory", Mode: 0o755},
		{Path: "docs/deep/nested.txt", Kind: "file", Mode: 0o600, Size: 0},
		// "docs" and "docs/deep" are deliberately missing from the manifest.
		{Path: "link", Kind: "symlink", Mode: 0o777, LinkTarget: "README.md"},
	}
}

func TestLegacyBaseTreeStructure(t *testing.T) {
	tree, err := NewLegacyBaseTree(legacyManifestEntries())
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()

	root, err := tree.GetInode(ctx, RootIno)
	if err != nil {
		t.Fatal(err)
	}
	if root.Inode.Kind != FileKindDirectory {
		t.Fatalf("root kind %s", root.Inode.Kind)
	}

	// Explicit ino preserved exactly.
	entry, err := tree.Lookup(ctx, root.Ref, "src")
	if err != nil {
		t.Fatal(err)
	}
	srcDir, err := tree.GetInode(ctx, entry.Ino)
	if err != nil {
		t.Fatal(err)
	}
	mainEntry, err := tree.Lookup(ctx, srcDir.Ref, "main.go")
	if err != nil {
		t.Fatal(err)
	}
	if mainEntry.Ino != 77 {
		t.Fatalf("explicit ino not preserved: %d", mainEntry.Ino)
	}

	// Synthesized intermediate directories resolve.
	docsEntry, err := tree.Lookup(ctx, root.Ref, "docs")
	if err != nil {
		t.Fatal(err)
	}
	if docsEntry.Kind != FileKindDirectory {
		t.Fatalf("docs kind %v", docsEntry.Kind)
	}
	docs, err := tree.GetInode(ctx, docsEntry.Ino)
	if err != nil {
		t.Fatal(err)
	}
	deepEntry, err := tree.Lookup(ctx, docs.Ref, "deep")
	if err != nil {
		t.Fatal(err)
	}
	deep, err := tree.GetInode(ctx, deepEntry.Ino)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := tree.Lookup(ctx, deep.Ref, "nested.txt"); err != nil {
		t.Fatal(err)
	}

	// Synthetic inos start above the maximum explicit ino (77).
	if docsEntry.Ino <= 77 {
		t.Fatalf("synthetic ino %d not above explicit max", docsEntry.Ino)
	}

	// ReadDir pages in name-byte order.
	var names []string
	cursor := ""
	for {
		entries, next, err := tree.ReadDir(ctx, root.Ref, cursor, 2)
		if err != nil {
			t.Fatal(err)
		}
		for _, e := range entries {
			names = append(names, e.Name)
		}
		if next == "" {
			break
		}
		cursor = next
	}
	want := []string{"README.md", "docs", "link", "src"}
	if len(names) != len(want) {
		t.Fatalf("names %v", names)
	}
	for i := range want {
		if names[i] != want[i] {
			t.Fatalf("names %v", names)
		}
	}

	// Symlink metadata.
	linkEntry, err := tree.Lookup(ctx, root.Ref, "link")
	if err != nil {
		t.Fatal(err)
	}
	link, err := tree.GetInode(ctx, linkEntry.Ino)
	if err != nil {
		t.Fatal(err)
	}
	if link.Inode.SymlinkTarget != "README.md" || link.Inode.Size != uint64(len("README.md")) {
		t.Fatalf("link inode %+v", link.Inode)
	}

	if _, err := tree.Lookup(ctx, root.Ref, "absent"); !errors.Is(err, ErrNotFound) {
		t.Fatalf("absent lookup: %v", err)
	}
	if _, err := tree.GetInode(ctx, 123456); !errors.Is(err, ErrNotFound) {
		t.Fatalf("absent ino: %v", err)
	}
}

func TestLegacyBaseTreeExtents(t *testing.T) {
	tree, err := NewLegacyBaseTree(legacyManifestEntries())
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()

	main, err := tree.GetInode(ctx, 77)
	if err != nil {
		t.Fatal(err)
	}

	t.Run("chunked file window spans chunk boundary", func(t *testing.T) {
		extents, err := tree.ReadExtents(ctx, main.Ref, 4194304-100, 200)
		if err != nil {
			t.Fatal(err)
		}
		if len(extents) != 2 {
			t.Fatalf("extents %+v", extents)
		}
		first, second := extents[0], extents[1]
		if first.Legacy == nil || first.Legacy.ObjectDigest != "sha256:c1" ||
			first.FileOffset != 4194304-100 || first.Length != 100 ||
			first.Legacy.ObjectOffset != 4194304-100 {
			t.Fatalf("first extent %+v %+v", first, first.Legacy)
		}
		if second.Legacy.ObjectDigest != "sha256:c2" || second.FileOffset != 4194304 ||
			second.Length != 100 || second.Legacy.ObjectOffset != 0 {
			t.Fatalf("second extent %+v %+v", second, second.Legacy)
		}
	})

	t.Run("whole-blob file", func(t *testing.T) {
		root, err := tree.GetInode(ctx, RootIno)
		if err != nil {
			t.Fatal(err)
		}
		readmeEntry, err := tree.Lookup(ctx, root.Ref, "README.md")
		if err != nil {
			t.Fatal(err)
		}
		readme, err := tree.GetInode(ctx, readmeEntry.Ino)
		if err != nil {
			t.Fatal(err)
		}
		extents, err := tree.ReadExtents(ctx, readme.Ref, 1, 100)
		if err != nil {
			t.Fatal(err)
		}
		if len(extents) != 1 {
			t.Fatalf("extents %+v", extents)
		}
		e := extents[0]
		if e.FileOffset != 1 || e.Length != 4 || e.Legacy.ObjectOffset != 1 || e.Legacy.ObjectSize != 5 {
			t.Fatalf("extent %+v %+v", e, e.Legacy)
		}
	})

	t.Run("empty file and past-EOF windows", func(t *testing.T) {
		root, _ := tree.GetInode(ctx, RootIno)
		docsEntry, _ := tree.Lookup(ctx, root.Ref, "docs")
		docs, _ := tree.GetInode(ctx, docsEntry.Ino)
		deepEntry, _ := tree.Lookup(ctx, docs.Ref, "deep")
		deep, _ := tree.GetInode(ctx, deepEntry.Ino)
		nestedEntry, err := tree.Lookup(ctx, deep.Ref, "nested.txt")
		if err != nil {
			t.Fatal(err)
		}
		nested, _ := tree.GetInode(ctx, nestedEntry.Ino)
		if extents, err := tree.ReadExtents(ctx, nested.Ref, 0, 10); err != nil || len(extents) != 0 {
			t.Fatalf("%v %v", extents, err)
		}
		if extents, err := tree.ReadExtents(ctx, main.Ref, 9000000, 10); err != nil || len(extents) != 0 {
			t.Fatalf("%v %v", extents, err)
		}
	})

	t.Run("directory rejected", func(t *testing.T) {
		root, _ := tree.GetInode(ctx, RootIno)
		if _, err := tree.ReadExtents(ctx, root.Ref, 0, 10); !errors.Is(err, ErrCorrupt) {
			t.Fatalf("got %v", err)
		}
	})
}

func TestLegacyBaseTreeRejectsBadManifests(t *testing.T) {
	cases := map[string][]LegacyEntry{
		"duplicate ino": {
			{Path: "a", Kind: "file", Mode: 0o644, Ino: 5},
			{Path: "b", Kind: "file", Mode: 0o644, Ino: 5},
		},
		"root ino claimed": {
			{Path: "a", Kind: "file", Mode: 0o644, Ino: 1},
		},
		"duplicate path": {
			{Path: "a", Kind: "file", Mode: 0o644},
			{Path: "a", Kind: "directory", Mode: 0o755},
		},
		"file as parent": {
			{Path: "a", Kind: "file", Mode: 0o644},
			{Path: "a/b", Kind: "file", Mode: 0o644},
		},
		"blob size mismatch": {
			{Path: "a", Kind: "file", Mode: 0o644, Size: 10, BlobDigest: "sha256:x", BlobSize: 9},
		},
		"missing blob": {
			{Path: "a", Kind: "file", Mode: 0o644, Size: 10},
		},
		"non-contiguous chunks": {
			{Path: "a", Kind: "file", Mode: 0o644, Size: 10, Chunks: []LegacyChunk{
				{Digest: "sha256:c", Size: 5, Offset: 1},
			}},
		},
		"chunk sum mismatch": {
			{Path: "a", Kind: "file", Mode: 0o644, Size: 10, Chunks: []LegacyChunk{
				{Digest: "sha256:c", Size: 5, Offset: 0},
			}},
		},
		"bad kind": {
			{Path: "a", Kind: "device", Mode: 0o644},
		},
		"bad path": {
			{Path: "a//b", Kind: "file", Mode: 0o644},
		},
		"dotdot path": {
			{Path: "a/../b", Kind: "file", Mode: 0o644},
		},
		"negative size": {
			{Path: "a", Kind: "file", Mode: 0o644, Size: -1},
		},
		"empty symlink target": {
			{Path: "l", Kind: "symlink", Mode: 0o777, LinkTarget: ""},
		},
		"symlink target with NUL": {
			{Path: "l", Kind: "symlink", Mode: 0o777, LinkTarget: "a\x00b"},
		},
		"timestamp beyond bound": {
			{Path: "a", Kind: "file", Mode: 0o644, MtimeMs: 1 << 60},
		},
		"negative timestamp beyond bound": {
			{Path: "a", Kind: "file", Mode: 0o644, CtimeMs: -(1 << 60)},
		},
	}
	for name, entries := range cases {
		if _, err := NewLegacyBaseTree(entries); err == nil {
			t.Fatalf("%s: accepted", name)
		}
	}
}

func TestLegacyModeMaskedToPermissionBits(t *testing.T) {
	// Both adapters keep only the 0o7777 mode bits (type bits are carried by
	// the kind), mirroring the TypeScript adapter exactly.
	tree, err := NewLegacyBaseTree([]LegacyEntry{{Path: "a", Kind: "file", Mode: 0o100644}})
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()
	root, err := tree.GetInode(ctx, RootIno)
	if err != nil {
		t.Fatal(err)
	}
	entry, err := tree.Lookup(ctx, root.Ref, "a")
	if err != nil {
		t.Fatal(err)
	}
	file, err := tree.GetInode(ctx, entry.Ino)
	if err != nil {
		t.Fatal(err)
	}
	if file.Inode.Mode != 0o644 {
		t.Fatalf("mode %#o", file.Inode.Mode)
	}
}

func TestLegacyHandlesFailClosedAgainstRealStores(t *testing.T) {
	tree, err := NewLegacyBaseTree(legacyManifestEntries())
	if err != nil {
		t.Fatal(err)
	}
	root, err := tree.GetInode(context.Background(), RootIno)
	if err != nil {
		t.Fatal(err)
	}
	// A legacy handle has size 0: below MinNodeBytes, so a PFT2 reader
	// rejects it before ever fetching.
	if err := checkNodeRefBounds("probe", root.Ref); !errors.Is(err, ErrInvalidNode) {
		t.Fatalf("legacy handle passed node ref bounds: %v", err)
	}
}
